### DISCUSSION TITLE
Use ESPHome + LVGL instead of Nextion display

### DISCUSSION THREAD

**Discussion by nagyrobi** [dzurikmiroslav/esp32-evse#59] [Ideas]
TL;DR:
https://github.com/nagyrobi/esp32evse_esphome-lvgl
---
Here's my plan to use ESPHome + LVGL instead of Nextion display:

ESPHome is [adding support](https://deploy-preview-3678--esphome.netlify.app/components/lvgl) for the LVGL graphics library.
Several screens are supported but for this project a WT32-SC01 Plus or Guition JC3248w535 would be most suitable because:
- has a pretty nice case that can be mounted on a box which has `esp32-evse` inside
- has resolution of 480*320
- has capacitive touch screen
- it runs an ESP32-S3
- has quite some free GPIOs
- A LOT less expensive than Nextion

Running ESPHome on it has the following advantages:
- any useful graphical UI can be rendered via LVGL, with fully open source resources (no proprietary editor needed)
- multiple UARTs can be configured (eg one talking to `esp32-evse`, other talking to smart meter etc)
- can connect to Home Assistant via native API (or MQTT to other systems), can expose any register as a sensor or control to HA
- can use free GPIOs or the display for anything else
- has Bluetooth and with it you could pair with your car and implement additional related features (eg open charging flap automatically, poll battery status, temperature; garage door control etc, which wouldn't have to interfere with EVSE board at all)

With this it becomes possible not only to have a fully customizable UI, but also:
- 1-click integration with Home Assistant, making possible any notifications or remote access through it
- make own Dynamic Load Management through HA or other ESPHome components (eg use data from other meters in various ways; freely implement DLM logic either in HA or in ESPHome)
- implement any custom authorization like RFiD readers etc

As far as I see the ModBUS feature set is wider that Nextion's so it's already possible to implement more functionality with this kind of display that with Nextion out of the box.

As soon as I get my `esp32s2-evse-d-a` up and running, will start on this and publish the YAML.

**Comment by dzurikmiroslav** [comment]
Yes, Nextion is quite outdated and their designer is from the Windows XP era :-) This device with LVGL look good :+1: 

With a Modbus API, the display must periodically pooling registers ​​from the EVSE. Probably in small time interval, which can be reflected in the performance.

Nextion API is designed to you subscribe to value, and EVSE send you once immediately and next time, when its changed value. I don't know if this data format would be suitable for LVGL, but it is not a problem to implement all other values like Modbus API.

Also you can use Netion API thorught RS485 serial, the same also applies to Modbus API, it can also be used via UART serial.

**Reply by nagyrobi** [reply]
Poll time can be configured at will.
Nextion is a proprietary format, ESPHome can also only send data but not receive aka. emulate a Nextion display. But, could be done in theory, if we had the protocol description.

**Reply by dzurikmiroslav** [reply]
It's a text format and not very complicated, some small example:

- display send: `sub state`
- EVSE send: `state.txt="A"`
- when EV connect EVSE send:  `state.txt="B0"`

**Reply by nagyrobi** [reply]
Actually the subscribe-like approach of the Nextion display here is killer, looked a bit closer and compared with ModBUS and I admit it's indeed more suitable. A bit more work but wort to be done.

**Reply by nagyrobi** [reply]
> It's a text format and not very complicated, some small example:
> 
>     * display send: `sub state`
>     * EVSE send: `state.txt="A"`
>     * when EV connect EVSE send:  `state.txt="B0"`

What happens when EVSE reboots, but Nextion display doesn't? Display sent `sub state` earlier, expects data but EVSE rebooted meanwhile `sub state` not set?

**Reply by nagyrobi** [reply]
I found that at EVSE reboot I get `rest\xFF\xFF\xFFsleep=0\xFF\xFF\xFF` from EVSE, are these intended for this?
Don't see `rest` or `sleep=0` in the nextion page.

**Reply by dzurikmiroslav** [reply]
It`s Nextion stuff https://nextion.tech/instruction-set/
I have no experience with LVGL displays yet, wouldn't it be better to implement own communication protocol in script?
https://github.com/dzurikmiroslav/esp32-evse/wiki/Lua#serial-module

**Reply by nagyrobi** [reply]
I'm almost done, when I finish I'll share the solution. It's pretty simple.

**Reply by nagyrobi** [reply]
The nextion-like protocol implemented, I can send commands and read statuses from ESPHome, and they appear in Home Assistant too.
GUI mockup in LVGL:
![kép](https://github.com/user-attachments/assets/3d033ea2-4e72-4dd5-b3db-693dc1e6bf83)
To share with the project, I'll reimplement your Nextion screens here.

**Reply by nagyrobi** [reply]
![kép](https://github.com/user-attachments/assets/419f72be-b32a-450a-b0c8-b3e3e0d6aff1)
Guition JC3248w535, direct conection UART <-> UART (3.3V), +5V and GND.

**Reply by dzurikmiroslav** [reply]
Good work :) white board look nice 👍 I see you not need socket lock, I made one board with 15W power supply (not 20W) is also suitable.

**Reply by nagyrobi** [reply]
Is the 5V output line powerful enough to hold the ESP32-S3 screen too? Might also add some peripherals to auto-detect plug removal from holder, to transmit 433MHz bits to open charging door flap.

**Reply by dzurikmiroslav** [reply]
The K7805M-1000R3 should give max 1A current.
Yes, this would be an interesting feature. But I don't own a Tesla :) I have nowhere to tune it.

**Reply by nagyrobi** [reply]
Can you add some more data to the Nextion API? I see it has less than ModBUS. What I specifically would like to access from the screen are:
- Charger available setting (1/0)
- Temperature threshold setting
- Default charging current setting
- Default consumption limit setting
- Default charging time limit setting
- Default underpower limit setting
- Low temperature value (place a sensor close to the floor to detect freeze in winter)
- Restart (implement like Authorize) to reboot EVSE on demand

**Reply by nagyrobi** [reply]
Today's result:
![kép](https://github.com/user-attachments/assets/8bbefb89-e95d-45ba-b44f-6c131378f436) ![kép](https://github.com/user-attachments/assets/522cb461-5238-4cee-8f5c-2b633e5f83d2)
![kép](https://github.com/user-attachments/assets/73b6680e-c6e3-4056-8a71-d77aa76509bf) ![kép](https://github.com/user-attachments/assets/c17c61ee-b0b8-4c5d-9a24-ce6e74dda6c2)

**Reply by nagyrobi** [reply]
Also did you implement authorize on the Nextion display? I could add a popup with a button just to demonstrate the authorization workflow.

**Reply by dzurikmiroslav** [reply]
Yes there is command `pendAuth` and `auth` https://github.com/dzurikmiroslav/esp32-evse/wiki/Nextion#table-of-commands

**Reply by nagyrobi** [reply]
I know about these but did you implement them in your Nextion display? Can't test as I don't have one and don't see in your screenshots

**Reply by nagyrobi** [reply]
Also do you know what's the Nextion variable name for Charging time? Only the limit is documented

**Reply by nagyrobi** [reply]
> Yes there is command `pendAuth` and `auth` https://github.com/dzurikmiroslav/esp32-evse/wiki/Nextion#table-of-commands

![kép](https://github.com/user-attachments/assets/aee73983-a6fa-4510-ba5a-0655f0c4a6d4)

**Reply by dzurikmiroslav** [reply]
> I know about these but did you implement them in your Nextion display? Can't test as I don't have one and don't see in your screenshots
From simulator:
![image](https://github.com/user-attachments/assets/f52bddff-a81c-440a-b2a6-239361d0cdb0)

**Reply by dzurikmiroslav** [reply]
> Also do you know what's the Nextion variable name for Charging time? Only the limit is documented

It is `chTime` I forgot to mention in wiki doc, added...

**Reply by nagyrobi** [reply]
Published: https://github.com/nagyrobi/esp32-evse_esphome-lvgl

**Reply by nagyrobi** [reply]
Updated

**Comment by Radwan-Alalili** [comment]
Is it okay if I use other HMI instead of using Nextion for example CrowPanel touch screen, does the program work fine or there will be some issues

**Reply by dzurikmiroslav** [reply]
I'm not familiar with CrowPanel, but I think any HMI with a UART or RS-485 interface is fine.

**Reply by Y00c** [reply]
Do I need to add additional code to  esp32  to connect the CrowPanel  screen?   
 Or do I just have to design the UI for the screen only and connect directly to the UART without modifying the main code of the ESP32 ?

**Reply by dzurikmiroslav** [reply]
No, just setup serial port mode to "Script" value. And from Lua script you have access to it https://github.com/dzurikmiroslav/esp32-evse/wiki/Lua#serial-module

**Reply by Y00c** [reply]
can you help  me to do that ?

**Reply by Y00c** [reply]
are you have  any  account in Social media  to contact you

**Reply by Y00c** [reply]
Hi

**Reply by dzurikmiroslav** [reply]
Hi, Sorry, but I'm too busy to solve individual tasks. If you need advice, please post it on the GitHub discussion. Then others who are solving a similar problem will find a solution.

**Reply by Y00c** [reply]
> Hi, Sorry, but I'm too busy to solve individual tasks. If you need advice, please post it on the GitHub discussion. Then others who are solving a similar problem will find a solution.
   Thank you dear

**Reply by Y00c** [reply]
> Hi, Sorry, but I'm too busy to solve individual tasks. If you need advice, please post it on the GitHub discussion. Then others who are solving a similar problem will find a solution.

   Could you please help me understand how to do this using Arduino IDE  with an LVGL  esp 32 display? Is there an example available? I couldn't implement it.

**Reply by Y00c** [reply]
> > Hi, Sorry, but I'm too busy to solve individual tasks. If you need advice, please post it on the GitHub discussion. Then others who are solving a similar problem will find a solution.
> 
> Could you please help me understand how to do this using Arduino IDE with an LVGL esp 32 display? Is there an example available? I couldn't implement it.
![20250316_232952](https://github.com/user-attachments/assets/6667bb5c-a92e-4a42-85c5-151d3de19955)

![20250316_233001](https://github.com/user-attachments/assets/7f4e3c5c-0b8e-4a02-adbc-d941edcf9520)
![20250316_233129](https://github.com/user-attachments/assets/7da059c5-9062-43ee-b4ed-28d1168d0228)
![20250316_233134](https://github.com/user-attachments/assets/48495e5c-420e-48e4-ba29-f
![20250316_233142](https://github.com/user-attachments/assets/cfdf3fb5-e807-41de-8679-8b3d041118a5)
6484cf91e4d)

**Reply by Y00c** [reply]
> Hi, Sorry, but I'm too busy to solve individual tasks. If you need advice, please post it on the GitHub discussion. Then others who are solving a similar problem will find a solution.

![20250316_233134](https://github.com/user-attachments/assets/8dd1900b-89e9-4089-8597-304343c98ed5)

**Comment by nagyrobi** [comment]
I thought of having a more precise description on the user interface for the states. What I know so far is:

- A: EV not connected
- B1: EV connected
- B2: EV connected
- C1: Charging
- C2: Charging
- D1: Charging with ventilation
- D2: Charging with ventilation
- E: Connection error
- F: Error

Could you help me understand better the difference between B1-B2, C1-C2 and D1-D2 respectively?

**Reply by nagyrobi** [reply]
Found [this](https://de.mathworks.com/help/autoblks/ug/charge-an-electric-vehicle.html), but I think it doesn't match exactly your implementation.
And [here](https://morth.gov.in/sites/default/files/ASI/320201825106PMAIS_138_Part_2_F.pdf) they list a plethora of other possible intermediate stated, but it's apparently for DC chargers...

**Reply by dzurikmiroslav** [reply]
B1 - is pre authorization state
B2 - is after authorization state
C1/D1 - EV want charging, but AC contractor not connected (for example paused charging, limit reached)
C2/D2 - EV want charging and AC contractor is on

I find out mistake in my diagram (twice has C state) here is fixed
![evse_state_map drawio](https://github.com/user-attachments/assets/7e19f9e1-dd2c-4583-8680-920890114355)

**Reply by nagyrobi** [reply]
Interesting, it's good we're talking about this. See my feature request [issue](https://github.com/dzurikmiroslav/esp32-evse/issues/113) for details as it appears to be a little different with my car.

**Reply by nagyrobi** [reply]
Corrected the textual state display in the ESPHome HMI yaml to:

- A: EV not connected
- B1: EV connected
- B2: EV connected, authorized
- C1: Ready to charge
- C2: Charging
- D1: Ready to charge with ventilation
- D2: Charging with ventilation
- E: Connection error
- F: Error
      
Hope it's correct.

**Comment by nagyrobi** [comment]
<img width="482" height="320" alt="kép" src="https://github.com/user-attachments/assets/f5990cf9-9a0b-4eaf-8691-8e6353791f78" />

**Comment by nagyrobi** [comment]
New version, for ESP32-EVSE v2.0 and up, based on AT commands:
https://github.com/nagyrobi/esp32evse_esphome-lvgl
Relies on ESPHome external component:
https://github.com/nagyrobi/esp32evse-esphome
Works on highter baudrates.

Config examples included for:
- Wireless-Tag/Panlee WT32-SC01 Plus
- Guition JC3248w535
- Guition 4848s040:
<img width="480" height="480" alt="kép" src="https://github.com/user-attachments/assets/dca43d1c-bc70-4b14-ac9a-c6c85d33ed1e" />
